# **Asynchronous programming: futures, async, await**

https://dart.dev/libraries/async/async-await#why-asynchronous-code-matters

## 동기(Synchronous), 비동기(Asynchronous)

**Key terms**

- **동기 작업(synchronous operation)**: 동기 작업은 작업이 완료될 때까지 다른 작업의 실행을 차단한다.
- **동기 함수(synchronous function)**: 동기 함수는 동기 작업만 수행한다.
- **비동기 작업(asynchronous operation)**: 비동기 작업은 해당 비동기 작업이 완료되기 전에 다른 작업을 실행할 수 있다.
- **비동기 함수(asynchronous function)**: 비동기 함수는 하나 이상의 비동기 작업을 수행하며 동기 작업도 포함될 수 있다.

## 비동기 코드는 왜 중요한가?

작업 소요 시간을 단축시킬 수 있다. 특정 작업이 완료되기까지 기다리지 않고 다른 작업을 수행할 수 있기 때문이다.

### 일반적인 비동기 작업

- 네트워크를 통한 데이터 송수신
- 데이터베이스 작업
- 파일 작업

### 비동기 작업의 결과

- **일회성** 결과는 **Future**를 제공
- **다회성** 결과는 **Stream**을 제공

Future를 반환하면 비동기 함수다. 비동기 작업을 수용하려면 다른 일반 함수도 비동기화되어야 한다. 비동기 결과와 상호 작용하려면 async와 await 키워드를 사용하면 된다.

### 비동기 함수를 잘못 사용한 경우

createOrderMessage()는 비동기 함수인 fetchUserOrder()가 완료될 때까지 기다리지 않기 때문에 fetchUserOrder()가 제공하는 문자열을 얻을 수 없다.

createOrderMessage()는 완료되지 않은 future를 얻는다.

```dart
String createOrderMessage() {
  var order = fetchUserOrder();
  return 'Your order is: $order';
}

Future<String> fetchUserOrder() =>
    // Imagine that this function is more complex and slow.
    Future.delayed(
      const Duration(seconds: 2),
      () => 'Large Latte',
    );

void main() {
  print(createOrderMessage());
}

// Result:
Your order is: Instance of '_Future<String>'
```

## Future란 무엇인가?

future는 비동기 작업의 결과를 나타내며, 완료(completed) 또는 미완료(uncompleted) 상태를 갖는다. ‘f’uture는 ‘F’uture 클래스의 인스턴스다.

### Uncompleted

Future를 반환하는 비동기 함수를 호출하면 수행할 작업을 큐에 추가하고 uncompleted future를 반환한다. 해당 future는 비동기 작업이 정상적으로 끝나거나 에러가 발생할 때까지 대기한다.

### Completed

비동기 작업이 성공적으로 끝나면 future는 value로 completed되고, 그렇지 않으면 error로 completed된다.

**Completing with a ‘value’**

Future<T> 타입의 future는 T 타입의 값으로 완료된다. Future<String> 타입의 future는 문자열 값을 생성한다. Future<void> 타입의 future는 사용 가능한 값을 생성하지 않는다.

**Completing with an ‘error’**

어떠한 이유로 비동기 작업이 실패하면 future는 에러로 완료된다.

print()의 문자열이 먼저 출력되고 fetchUserOrder()가 오류와 함께 완료된다.

```dart
Future<void> fetchUserOrder() {
  // Imagine that this function is fetching user info but encounters a bug.
  return Future.delayed(
    const Duration(seconds: 2),
    () => throw Exception('Logout failed: user ID is invalid'),
  );
}

void main() {
  fetchUserOrder();
  print('Fetching user order...');
}

// Result:
Fetching user order...
DartPad caught unhandled _Exception:
Exception: Logout failed: user ID is invalid
https://storage.googleapis.com/nnbd_artifacts/3.4.3/dart_sdk.js 11994:11  Object.throw_
blob:null/33a2d4ac-5044-4554-83d7-cdab85d97d78 382:76                     <fn>
https://storage.googleapis.com/nnbd_artifacts/3.4.3/dart_sdk.js 41793:33  <fn>
...
https://storage.googleapis.com/nnbd_artifacts/3.4.3/dart_sdk.js 47652:18  Object.runZoned
blob:null/33a2d4ac-5044-4554-83d7-cdab85d97d78 417:20                     Chain.capture
Stack trace truncated and adjusted by DartPad...
```

## Async와 Await

async, await 키워드는 비동기 함수를 정의하고 비동기 함수의 결과를 사용하기 위해 결과를 얻는 선언적 방법을 제공한다.

### async와 await 키워드 사용 기본 지침

- 비동기 함수를 정의하려면 함수 본문 앞에 async를 추가한다.
- await는 async 함수에서만 사용 가능하다.

```dart
void main() async {
	1();
	await 2();
	3();
	await 4();
}
```

```dart
void main() async { ··· }
```

```dart
Future<void> main() async { ··· }
```

await를 사용하면 future가 완료될 때까지 대기한다.

```dart
print(await createOrderMessage());
```

### async와 await 키워드를 언제 사용해야 하는가?

비동기 작업의 결과를 동기 작업을 실행했을 때처럼 async 함수 내에서 절차에 맞게 결과를 얻어 사용하기 위해 await로 대기시킨다. 그러나 다른 모든 함수의 실행을 차단하는 것은 아니다. await는 async가 붙은 함수 내에서만 사용 가능하다. 동기 코드와 유사한 비동기 코드가 생성되는 것이다.

```dart
Future<String> createOrderMessage() async {
  var order = await fetchUserOrder();
  return 'Your order is: $order';
}

Future<String> fetchUserOrder() =>
    // Imagine that this function is
    // more complex and slow.
    Future.delayed(
      const Duration(seconds: 2),
      () => 'Large Latte',
    );

Future<void> main() async {
  print('Fetching user order...');
  print(await createOrderMessage());
}

// Result:
Fetching user order...
Your order is: Large Latte
```

### async와 await 키워드를 사용하면 실행 순서에 어떤 영향을 주는가?

await를 사용한다고 해서 다른 모든 함수의 실행이 차단되는 것은 아니다. await를 사용한 async 함수 내에서만 비동기 작업 결과를 대기하는 것이다. 이를 통해 비동기적으로 작업을 수행한다.

비동기 함수는 첫 번째 await 키워드가 나오기 전까지 동기적으로 실행된다. 비동기 함수 본문 내에서 첫 번째 await 키워드 이전의 모든 코드가 즉시 실행된다는 것을 의미한다.

await로 대기시키지 않으면 동기식 코드처럼 실행 흐름이 그대로 흘러가서 비동기 작업 결과가 완성되기 전에 프로그램이 종료될 수 있다.

## Error Handling

비동기 함수의 에러를 처리하기 위해 다음과 같이 try-catch를 사용할 수 있다. 동기 코드와 동일한 방식이다.

```dart
try {
  print('Awaiting user order...');
  var order = await fetchUserOrder();
} catch (err) {
  print('Caught error: $err');
}
```

```dart
Future<void> printOrderMessage() async {
  try {
    print('Awaiting user order...');
    var order = await fetchUserOrder();
    print(order);
  } catch (err) {
    print('Caught error: $err');
  }
}

Future<String> fetchUserOrder() {
  // Imagine that this function is more complex.
  var str = Future.delayed(
      const Duration(seconds: 4),
      () => throw 'Cannot locate user order');
  return str;
}

void main() async {
  await printOrderMessage();
}

// Result:
Awaiting user order...
Caught error: Cannot locate user order
```
